picky-asn1-x509/src/pkcs12/authenticated_safe.rs

```diff
@@ -0,0 +1,102 @@
+use crate::{oids, pkcs12::SafeContentsContentInfo};
+use core::fmt;
+use oid::ObjectIdentifier;
+use picky_asn1::wrapper::{Asn1SequenceOf, ExplicitContextTag0, ObjectIdentifierAsn1, OctetStringAsn1Container};
+use picky_asn1_der::Asn1RawDer;
+use serde::{de, ser};
+
+/// Top-level `ContentInfo` type used in context of `AuthenticatedSafe` for PKCS#12 `PFX` structure.
+/// Defined in [PKCS #12](https://datatracker.ietf.org/doc/html/rfc7292#section-3.4)
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum AuthenticatedSafeContentInfo {
+    Data(Vec<SafeContentsContentInfo>),
+    /// Unknown `ContentInfo` type used in context of `AuthenticatedSafe`.
+    /// Most likely `SignedData` is used, which is not currently supported by picky.
+    Unknown {
+        content_type: ObjectIdentifier,
+        content: Option<Asn1RawDer>,
+    },
+}
+
+impl ser::Serialize for AuthenticatedSafeContentInfo {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: ser::Serializer,
+    {
+        use ser::SerializeSeq;
+
+        match self {
+            AuthenticatedSafeContentInfo::Data(safe_contents) => {
+                let mut seq = serializer.serialize_seq(Some(2))?;
+                let oid: ObjectIdentifierAsn1 = oids::content_info_type_data().into();
+                seq.serialize_element(&oid)?;
+                let content: ExplicitContextTag0<OctetStringAsn1Container<Asn1SequenceOf<SafeContentsContentInfo>>> =
+                    OctetStringAsn1Container(Asn1SequenceOf(safe_contents.clone())).into();
+                seq.serialize_element(&content)?;
+                seq.end()
+            }
+            AuthenticatedSafeContentInfo::Unknown { content_type, content } => {
+                let sequence_length = 1 + content.is_some() as usize;
```

Comment:
Nit: it is generally well-advised to not use "as" casting in most cases.

`From` is only implemented when the conversion is lossless and infallible. It’s incredibly useful to have a compile-time error when code is changed later.

When conversion is fallible, let’s use `TryFrom`, it’s fine to just `usize::try_from(v).unwrap()` in most cases. This way, it’s simpler to audit fallible conversions using `grep` and it’s easier to debug (we’ll get an immediate crash along a stacktrace when the value is invalid). Of course, in production code, we may want to handle the error in a more elegant manner, but if we can be sure it never fails, it’s not strictly necessary (in this case, the conversation can never fail because `usize` can always hold values 0 and 1).

Note that `usize::try_from(v).unwrap()` does not incur any performance penalty when redundant. For instance, casting a u32 to usize is actually infallible on 32 and 64 bits architectures and code is optimized accordingly: https://gcc.godbolt.org/z/Y4WjWTqh1
On the other hand, we actually want to check when this property does not actually hold.

For reference, I consider using `as` to be appropriate when:
- Using it as a building block to implement `From` or `Into`
- Using it as a building block to implement `as_X` methods

Clippy lint about this: https://rust-lang.github.io/rust-clippy/master/#cast_lossless

Nit 2: this pattern is not very frequent, and it took me some time to realize this is basically equivalent to the following:
```rust
let sequence_length = content.is_some() {
    2
} else {
    1
};
```
However, this sounds useful when there are a lot of optional elements. If it appears at other places, it’s good to stick with the `usize::from` conversion instead of the `if … else …` shape.

Reply:
`bool` -> `usize` conversion is infallible and well-defined (`false` -> `0`, `true` -> `1`) and in that case, it is just more compact alternative to `usize::from(val)`, and it has relatively the same level of readability as `usize::from`. This pattern is used in lot of places for optional values in `pkcs12` code, and it is much more compact than `if ... {} else ...`, especially in cases where there is more than one optional value:
```
let sequence_size = 2 + self.key_length.is_some() as usize + self.prf.is_some() as usize;
```